design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+   create a `Deployment`).
+7. Wait for `PackageRevision` to become `Ready` (this will occur after
+   successful install of all objects included in the `PackageRevision`).
+8. Set the `status.currentHash` of the `Package` to match the `sha256` hash of
+   the package image and `status.currentImage` to match the package image name.
+
+
+*Deleting a `Package`*
+
+1. Observe delete of a `Package`.
+2. Get the `Active` `PackageRevision` and check that there are no other `Active`
+   package revisions (for other `Package`s) that depend on CRDs that are managed
+   by this `Package` and mark as `Stopping`.
+3. When the `Stopping` `PackageRevision` has finished cleaning up its controller
+   it should be marked as `Inactive`. and the _`Package` controller_ should
+   clean up any CRDs, `InfrastructureDefinition`s, `InfrastructurePublication`s,
+   and `Composition`s that have a label corresponding to the `Package`.
+4. All `PackageRevision`s that are labeled with `Package` name should be
+   deleted.
+
+*Optional Features*
+
+The following are optional features that could be implemented in a future
+iterations:
+
+- Block deletion until all CRDs are manually cleaned up.
+- Only retain a specified number of `Inactive` `PackageRevision`s. This could be
+  accomplished by adding a field to the `Package` `spec` such as
+  `revisionHistoryLimit`. The _`Package` controller_ would then clean up any
+  `PackageRevision`s that are not within the `revisionHistoryLimit`.
+
+#### PackageRevision Controller
+
+Once a `PackageRevision` has been created, the _`Package` controller_ hands
+control to the _`PackageRevision` controller_. The _`PackageRevision`
+controller_ will operate similarly to current implementation the _`Package`
+controller_, but will also handle much of the functionality currently handled by
+the existing _`Package` controller_. In total, the _`PackageRevision`_
+controller will execute the following steps:
+
+*`PackageRevision` Created with State `Pending`*
+
+1. Observe creation of a `PackageRevision`.
+2. Wait until the `spec.state` is marked as `Active`.
+3. Read the logs of the install `Job` and parsing all CRDs,
+   `InfrastructureDefinition`s, `InfrastructurePublication`s, and
+   `Composition`s. The install `Job` will be able to be located by the
+   _`PackageRevision` controller_ because it will have the same name as the
+   parent `Package`.
+   - If the hash of the image in the `initContainer` of the install `Job` does
+     not match the name of the `PackageRevision`, do not proceed.
+4. Label and add UI metadata to CRDs and composition types.
+5. Ensure that all dependent (listed in `sped.dependsOn`) CRDs are present in
+   the cluster.
+6. Create or update all parsed objects from the `Job` output, and any RBAC roles
+   that are required. These will include the current `ClusterRole`s as well as
+   those that are required to give Crossplane the ability to manage the CRDs
```

Comment:
are these explicitly defined in the package, or does this controller automatically create them?

Reply:
@jbw976 currently, the package manager gives the controller the ability to create / update / delete / watch, etc. CRDs that it installs as well as any that it depends on. This stanza also mentions creating `ClusterRole`s for the composition controller to be able to manage CRDs created as a result of `InfraDef` / `InfraPub`. I believe we are now thinking that this will be handled by a separate RBAC controller, so the package manager will only be responsible for creating RBAC for the controller if one is being installed.